src/rust/engine/src/tasks.rs

```diff
@@ -45,43 +45,106 @@ impl rule_graph::Rule for Rule {
       &Rule::Intrinsic(_) => false,
     }
   }
+
+  fn color(&self) -> Option<rule_graph::Palette> {
+    match self {
+      Rule::Task(_) => None,
+      Rule::Intrinsic(_) => Some(rule_graph::Palette::Gray),
+    }
+  }
+
+  fn fmt_for_graph(&self) -> String {
+    match self {
+      Rule::Task(ref task) => {
+        let TaskElements {
+          rule_type,
+          task_name,
+          clause_portion,
+          product,
+          get_portion,
+        } = Self::extract_task_elements(task, 2, 1);
+
+        format!(
+          "@{}({}) -> {}{}\n{}",
+          rule_type, clause_portion, product, get_portion, task_name,
+        )
+      }
+      Rule::Intrinsic(ref intrinsic) => format!(
+        "@rule(<intrinsic>({}) -> {})",
+        intrinsic.input, intrinsic.product,
+      ),
+    }
+  }
+}
+
+struct TaskElements {
```

Comment:
Nice! I really like this generalization!

Consider renaming this to something like `FormattedTaskElements`, `FormattedTaskRuleElements`, or `FormattedRuleElements`. Something that explains that this abstraction is entirely to do with formatting.